Run inference in an Executor

mikaylagawarecki opened the pull request:
Experiment: run model predictions in the backend in a ThreadPoolExecutor so that each model prediction does not block reading requests from the queue

Baseline is reset in above PR that bugfixes a lot of the metrics calculations but I kept the metrics here anyway



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116190
* #116189
* #116188
* #116187
* __->__ #115286

mikaylagawarecki committed:
Run inference in an Executor

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Run inference in an Executor"

Experiment: run model predictions in the backend in an Executor so that each model prediction does not block reading requests from the queue

Generally this seemed to decrease the warmup latency but increase the average latency and decrease throughput, perhaps due to overhead from the Executor

Next step: Try ProcessPoolExecutor instead of ThreadPoolExecutor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Run inference in an Executor"

Experiment: run model predictions in the backend in an Executor so that each model prediction does not block reading requests from the queue

Generally this seemed to decrease the warmup latency but increase the average latency and decrease throughput, perhaps due to overhead from the Executor

Next step: Try ProcessPoolExecutor instead of ThreadPoolExecutor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Run inference in an Executor"

Experiment: run model predictions in the backend in an Executor so that each model prediction does not block reading requests from the queue

Generally this seemed to decrease the warmup latency but increase the average latency and decrease throughput, perhaps due to overhead from the Executor

Next step: Try ProcessPoolExecutor instead of ThreadPoolExecutor




[ghstack-poisoned]

albanD reviewed (commented):
Looks like the result is better with the thread pool?

mikaylagawarecki commented:
@albanD, not really only the warmup latency is consistently better. Average latency generally went up (especially as batch_size increased) and throughput was consistently worse

mikaylagawarecki committed:
Update on "Run inference in an Executor"

Experiment: run model predictions in the backend in an Executor so that each model prediction does not block reading requests from the queue

Generally this seemed to decrease the warmup latency but increase the average latency and decrease throughput, perhaps due to overhead from the Executor

Next step: Try ProcessPoolExecutor instead of ThreadPoolExecutor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Run inference in an Executor"

Experiment: run model predictions in the backend in an Executor so that each model prediction does not block reading requests from the queue

Generally this seemed to decrease the warmup latency but increase the average latency and decrease throughput, perhaps due to overhead from the Executor

Next step: Try ProcessPoolExecutor instead of ThreadPoolExecutor




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Run inference in an Executor"

Experiment: run model predictions in the backend in an Executor so that each model prediction does not block reading requests from the queue






[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Run inference in an Executor"

Experiment: run model predictions in the backend in a ThreadPoolExecutor so that each model prediction does not block reading requests from the queue

Baseline is reset in above PR that bugfixes a lot of the metrics calculations but I kept the metrics here anyway






[ghstack-poisoned]

albanD reviewed (approved):
SGTM
